### ISSUE TITLE
Cannot read properties of undefined (reading 'replicas')  [calling "get"]

### ISSUE THREAD

**Issue by danielbush** [redis/node-redis#2704]
### Description

Environment:

- client: node 18 (public.ecr.aws/docker/library/node:18-alpine)
- redis server: AWS elasticache
    - engine version: 6.2.6
    - 2 shards; 2 nodes per shard

We're getting this error - I think it's intermittent ie it works maybe even a lot of the time:

Stacktrace (redactions using `xxx`):

```
TypeError: Cannot read properties of undefined (reading 'replicas')
    at RedisClusterSlots.getSlotRandomNode (/server/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:118:19)
    at RedisClusterSlots.getClient (/server/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:110:37)
    at Commander._RedisCluster_execute (/server/node_modules/@redis/client/dist/lib/cluster/index.js:217:79)
    at Commander.sendCommand (/server/node_modules/@redis/client/dist/lib/cluster/index.js:129:98)
    at Commander.commandsExecutor (/server/node_modules/@redis/client/dist/lib/cluster/index.js:126:75)
    at BaseClass.<computed> [as get] (/server/node_modules/@redis/client/dist/lib/commander.js:8:29)
    at cacheItem (/server/node_modules/xxx:46:34)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async HttpClient.xxx (/server/node_modules/xxx:193:29)
    at async Promise.all (index 1)
    at async prefetchTask (/server/api/controllers/tasks.js:349:5)
    at async Promise.all (index 2)
```

You can see HttpClient (an internal library being asked to make a request.  It then calls `cacheItem` (an internal module) which then calls `redis.get` as per code below to check if there's something in the cache using `get` before making the http request.

Code:

We get the client like this (within an async function).  It is long-running, we keep using the same instance once created.  Note we pass only one url to `rootNodes` - the cluster configuration endpoint provided by AWS Elasticache.  (I wonder if that is an issue as the docs say to pass several nodes, I struggled to find docs on this).

```ts
    ...
    redisClient = redis.createCluster({
        // Assumes we are using AWS Elasticache cluster configuration endpoint
        // or similar.
        rootNodes: [{ url: config.url }],
        defaults: {
            // IMPORTANT: Simply not specifying `socket` here seems to
            // cause the client to hang regardless of what is in
            // `socket`.  An empty object suffices to make this go away.
            // It's mentioned here:
            // https://github.com/redis/node-redis/issues/1656#issuecomment-1212691760
            socket: {
                // If this is too small, we might get `Error: All the root nodes are unavailable`.
                connectTimeout: 5000,
                tls: true,
            },
        },
    });
    redisClient.on('error', error =>
        console.error(
            `getRedisClient: error occurred for ${config.url}`,
            error,
        ),
    );
    // Wait for it to connect to avoid any errors.
    await redisClient.connect();
    return redisClient;
    ...
```

Error is triggered in code like this using the above `redisClient`:

```ts
        const redis = await getRedisClient();
        try {
            cached = await redis.get(
                commandOptions({ returnBuffers: true }),
                key,
            );
        } catch (err) {
           ... log the error here ...
```

### Node.js Version

18.19.1

### Redis Server Version

6.2.6

### Node Redis Version

4.6.13

### Platform

Linux

**Comment by martinssonj** [comment]
We also get a similar problem sometimes when using cluster mode against elasticache 7.1

```
TypeError: Cannot read properties of undefined (reading 'master')
    at RedisClusterSlots.getClient (/usr/src/app/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:108:59)
    at Commander._RedisCluster_execute (/usr/src/app/node_modules/@redis/client/dist/lib/cluster/index.js:217:79)
    at Commander.sendCommand (/usr/src/app/node_modules/@redis/client/dist/lib/cluster/index.js:129:98)
    at Commander.commandsExecutor (/usr/src/app/node_modules/@redis/client/dist/lib/cluster/index.js:126:75) 
```

Happens when we do many set commands in parallel.

**Comment by MagnusHJensen** [comment]
We are facing a similar issue, also pointing to a serverless elasticache 7.1.

Is it something that we can fix/help on our end?

**Comment by danielbush** [comment]
I tried redis 4.7 - I think it includes https://github.com/redis/node-redis/pull/2731  - but I'm still getting the original issue.

**Comment by noor-siddique-paypay** [comment]
`rejectUnauthorized: false` might be helpful
```
    redisClient = redis.createCluster({
        rootNodes: [{ url: config.url }],
        defaults: {
            socket: {
                tls: true,
                rejectUnauthorized: false
            },
        },
    });
```

**Comment by edwin-jones** [comment]
> I tried redis 4.7 - I think it includes #2731 - but I'm still getting the original issue.
> 
> If anyone is seeing the contrary, let me know.

I think your change was reset in this merge: https://github.com/redis/node-redis/commit/b2d35c528654046eaed5d5d32099de32384400ae

I think you'll need to reapply your fix in new commits to fix this regression.

![image](https://github.com/user-attachments/assets/89494816-2ee0-448b-8c8e-e44cc8749089)

**Comment by gruppferi** [comment]
Hi
We are blocked with our migration due to this, any chance get fixed ?
Thanks in advance

**Comment by RobertHunter-Pluto** [comment]
I'm getting this as well. Which version was the fix originally merged into and then where did it regress, if it was prior to 4.7?

**Comment by dyoldie** [comment]
Hi, 
I am currently facing this issue as well. Would really appreciate if we could implement the fix. In the meantime if this is the only issue that is blocking people you can fix to version `4.6.14` which has the original fix implemented
Thanks

**Comment by dominique-claes** [comment]
Is there any progress when a new version will be released? We use "@keyv/redis": "^4.2.0", to connect to a Redis container in our K8S cluster. It gives the following error:

TypeError: Cannot read properties of undefined (reading 'replicas')
at RedisClusterSlots.getSlotRandomNode (/srv/node_modules/.pnpm/@redis+client@1.6.0/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:118:19)
at RedisClusterSlots.getClient (/srv/node_modules/.pnpm/@redis+client@1.6.0/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:110:37)
at Commander._RedisCluster_execute (/srv/node_modules/.pnpm/@redis+client@1.6.0/node_modules/@redis/client/dist/lib/cluster/index.js:217:79)
at Commander.sendCommand (/srv/node_modules/.pnpm/@redis+client@1.6.0/node_modules/@redis/client/dist/lib/cluster/index.js:129:98)
at Commander.commandsExecutor (/srv/node_modules/.pnpm/@redis+client@1.6.0/node_modules/@redis/client/dist/lib/cluster/index.js:126:75)
at Commander.BaseClass.<computed> [as get] (/srv/node_modules/.pnpm/@redis+client@1.6.0/node_modules/@redis/client/dist/lib/commander.js:8:29)
at KeyvRedis.get (file:///srv/node_modules/.pnpm/@keyv+redis@4.2.0/node_modules/@keyv/redis/dist/index.js:221:32)
at Keyv.get (file:///srv/node_modules/.pnpm/keyv@5.2.3/node_modules/keyv/dist/index.js:513:21)
at readData (file:///srv/build/server/index.js:1243:20)
at Object.getSession (/srv/node_modules/.pnpm/react-router@7.1.3_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/react-router/dist/development/index.js:9527:25)

The error occurs when we navigate to our web application after a user is logged on and their session needs to be stored in Redis. Our web application will show a general error page but after a refresh, everything works fine.

**Comment by stondini** [comment]
Hello

Exact same behavior.
Does the team plan a new patch ?

**Comment by pankajupadhyay29** [comment]
I encountered the same issue and, upon investigation, realised that when the same client tries to open multiple connections to the cluster, it fails. I was running this client on Kubernetes, and it worked fine initially, but I encountered an error when scaling up during periods of high load. During this load, the API received new requests before the connection was fully established, resulting in failures.

To address this, I modified the behaviour to perform a ping to the Redis cluster before starting the server. This change ensures that the connection is established before any requests are accepted, which resolves the issue.

**Comment by EricQi7** [comment]
**I'm still encountering the same issue with @redis/client@1.6.1. Has this been resolved in @redis/client@5.1.1?
My configuration is shown below:**
```
    redis: 
      cluster:
        rootNodes:
          - url: rediss://[clustercfg endpoint in AWS]:6379
        defaults:
          socket:
            tls: true
            rejectUnauthorized: false
```

`
techdocs error Request failed with status 500 Cannot read properties of undefined (reading 'replicas') type="errorHandler" stack="TypeError: Cannot read properties of undefined (reading 'replicas')\n at RedisClusterSlots.getSlotRandomNode (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:118:19)\n at RedisClusterSlots.getClient (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:110:37)\n at RedisCluster._RedisCluster_execute (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/index.js:217:79)\n at RedisCluster.sendCommand (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/index.js:129:98)\n at RedisCluster.commandsExecutor (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/index.js:126:75)\n at BaseClass. [as get] (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/commander.js:8:29)\n at KeyvRedis.get (/Users/Documents/Repository/node_modules/@backstage/backend-defaults/node_modules/@keyv/redis/dist/index.cjs:262:32)\n at async Keyv.get (/Users/Documents/Repository/node_modules/keyv/dist/index.cjs:478:21)\n at async DefaultCacheClient.get (/Users/Documents/Repository/node_modules/@backstage/backend-defaults/src/entrypoints/cache/CacheClient.ts:52:19)\n at async CachedEntityLoader.getFromCache (/Users/Documents/Repository/node_modules/@backstage/plugin-techdocs-backend/src/service/CachedEntityLoader.ts:63:13)\n at async CachedEntityLoader.load (/Users/Documents/Repository/node_modules/@backstage/plugin-techdocs-backend/src/service/CachedEntityLoader.ts:45:18)\n at async (/Users/Documents/Repository/node_modules/@backstage/plugin-techdocs-backend/src/service/router.ts:161:20)"
`

**Comment by nkaradzhov** [comment]
Hi, the project has moved significantly in the last couple of months. Can you confirm if the issue reproduces with the latest version?

**Comment by EricQi7** [comment]
Hi nkaradzhov. Our project uses [@keyv/redis](https://www.npmjs.com/package/@keyv/redis), which depends on the [node-redis](https://www.npmjs.com/package/@redis/client) package. Once @keyv/redis updates to the latest version of node-redis, I will attempt to reproduce this issue.

**Comment by yonran** [comment]
This error still happens in the latest 5.5.6 but I think it indicates a misuse of the @redis/client library. We were setting global `cluster = createCluster(); await cluster.connect()` but were making calls before `connect()` completed. We should have been setting the global variable only after `connect()` completed.

```
var {createCluster} = require('redis')
var host = 'localhost';
var cluster = createCluster({rootNodes: [{url: `redis://${host}:6379`}]}).on('error', err => console.log('Redis Cluster Error', err))
cluster.connect() // oops did not await the promise
await cluster.get('hi')

Uncaught TypeError: Cannot read properties of undefined (reading 'replicas')
```

Edit: As of 2 weeks since we upgraded to redis client 5 and made sure not to use the client until `connect()` resolves, we have not seen this message in our logs anymore (we used to get it every day or every other day).

**Comment by darcyrush** [comment]
We had the same issue as @yonran forgetting to connect first due to debugging other dependencies. 

@nkaradzhov perhaps it might be an idea to throw a more clear error if the get or set functions are called before a connection is complete?

**Comment by PaulAnnekov** [comment]
@EricQi7 here is the reason for the error when using keyv/redis: https://github.com/jaredwray/keyv/issues/1529. It looks like node-redis lacks support of Cluster status. "isOpen" doesn't mean the same for Cluster as for regular Client. And there is no "isReady" at all for Cluster. So currently there is no good property of Cluster which will answer the question "are we ready to execute commands?".

**Comment by metart43** [comment]
> This error still happens in the latest 5.5.6 but I think it indicates a misuse of the @redis/client library. We were setting global `cluster = createCluster(); await cluster.connect()` but were making calls before `connect()` completed. We should have been setting the global variable only after `connect()` completed.
> 
> ```
> var {createCluster} = require('redis')
> var host = 'localhost';
> var cluster = createCluster({rootNodes: [{url: `redis://${host}:6379`}]}).on('error', err => console.log('Redis Cluster Error', err))
> cluster.connect() // oops did not await the promise
> await cluster.get('hi')
> 
> Uncaught TypeError: Cannot read properties of undefined (reading 'replicas')
> ```
> 
> Edit: As of 2 weeks since we upgraded to redis client 5 and made sure not to use the client until `connect()` resolves, we have not seen this message in our logs anymore (we used to get it every day or every other day).

@yonran thank you so much for this! It helped to me address the issue. In my redis initialisation function I added the block to `await cluster.connect()` and the error disappeared!
